components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SAMLArtifactBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.sso.saml.builders;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.xml.security.utils.Base64;
+import org.joda.time.DateTime;
+import org.opensaml.saml2.core.Assertion;
+import org.wso2.carbon.identity.base.IdentityException;
+import org.wso2.carbon.identity.sso.saml.SAMLSSOConstants;
+import org.wso2.carbon.identity.sso.saml.builders.assertion.ExtendedDefaultAssertionBuilder;
+import org.wso2.carbon.identity.sso.saml.dao.SAML2ArtifactInfoDAO;
+import org.wso2.carbon.identity.sso.saml.dao.impl.SAML2ArtifactInfoDAOImpl;
+import org.wso2.carbon.identity.sso.saml.dto.SAML2ArtifactInfo;
+import org.wso2.carbon.identity.sso.saml.dto.SAMLSSOAuthnReqDTO;
+import org.wso2.carbon.identity.sso.saml.exception.ArtifactBindingException;
+import org.wso2.carbon.identity.sso.saml.util.SAMLSSOUtil;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+/**
+ * This class is used to build the saml2 artifact.
+ */
+public class SAMLArtifactBuilder {
+
+    private static final Log log = LogFactory.getLog(SAMLArtifactBuilder.class);
+
+    /**
+     * Build the SAML V2.0 Artifact type of Type Code 0x0004 and save it with SAML assertion, in the database.
+     * Artifact length : 44 bytes
+     * <p>
+     * SAML V2.0 defines an artifact type of type code 0x0004
+     * Identification:urn:oasis:names:tc:SAML:2.0:artifact-04
+     * <p>
+     * SAML_artifact := B64(TypeCode EndpointIndex RemainingArtifact)
+     * TypeCode := Byte1Byte2
+     * EndpointIndex := Byte1Byte2
+     * <p>
+     * TypeCode := 0x0004
+     * RemainingArtifact := SourceID MessageHandle
+     * SourceID := 20-byte_sequence
+     * MessageHandle := 20-byte_sequence
+     *
+     * @param authnReqDTO    SAML SSO authentication request.
+     * @param sessionIndexId Session index ID.
+     * @return SAML V2.0 Artifact, type of TypeCode 0x0004
+     */
+    public String buildSAML2Artifact(SAMLSSOAuthnReqDTO authnReqDTO, String sessionIndexId)
+            throws IdentityException, NoSuchAlgorithmException, ArtifactBindingException {
```

Comment:
Even here I don't think NoSuchAlgorithm should be thrown to the layer above.
IMO it should be handled within this implementation